File: ios/TappyAI/Core/Share/ShareArtifact.swift
import Foundation

/// The one canonical TappyAI share artifact — a port of `src/lib/share/shareArtifact.ts`
/// (and `ShareArtifact.kt` on Android).
///
/// 🔑 SAME INPUT, SAME TEXT AS THE WEB. The brochure format is deliberately identical to the
/// web builder for the same `PlacesLiveView`, so a recommendation shared from iOS and one
/// shared from the web look like they came from the same product — because they did.
///
/// 🚨 WHITELIST, NEVER BLACKLIST. `SharedPlace` names every field that may leave the app and
/// `pickPlace` copies exactly those. Provenance, evidence types, ranks, entity ids, `_tappy_*`
/// and conversation ids have no field to land in.
///
/// 🚨 NO `distanceKm`. It is a fact about the SENDER's position (privacy), so it is not in the
/// whitelist and cannot be shared.
struct ShareArtifact: Identifiable, Equatable, Sendable {
    enum Kind: Equatable, Sendable { case places, plan }

    let kind: Kind
    let title: String
    let subject: String
    let text: String
    /// The brand entry point — the only URL the share guard admits for a recommendation.
    let url: String
    let places: [SharedPlace]

    var id: String { subject + "\u{1F}" + String(text.hashValue) }
}

struct SharedPlace: Equatable, Sendable {
    let name: String
    var category: String? = nil
    var rating: Double? = nil
    var ratingCount: Int? = nil
    var address: String? = nil
    var phone: String? = nil
    var openingHours: String? = nil
    var priceRangeText: String? = nil
    var reasons: [String] = []
    var links: [SharedLink] = []
    var image: String? = nil
}

struct SharedLink: Equatable, Sendable {
    let kind: String
    let url: String
    var platform: String? = nil
}

enum ShareArtifactBuilder {
    /// chat_messages.body CHECK constraint on the web Inbox; text handoffs use the same bound.
    static let inboxMaxBody = 4000

    private static let sharedActionKinds: Set<String> = ["maps", "website", "review", "order", "booking", "ticket", "reservation"]

    private struct Labels {
        let recommends, plan, reviews, why, maps, website, review, order, booking, ticket, reservation, more, footer, people, budget: String
    }

    /// The brochure labels live in the String Catalog (`share.brochure.*`) and are resolved for
    /// the BROCHURE language — the recipient's — not the device locale. Their values are pinned
    /// equal to the web/Android tables by `crossPlatformShare.test.ts`, so the three clients
    /// keep producing the same text.
    private static func labels(_ lang: String) -> Labels {
        let locale = Locale(identifier: lang.hasPrefix("en") ? "en" : "vi")
        func s(_ key: String.LocalizationValue) -> String { String(localized: key, locale: locale) }
        return Labels(
            recommends: s("share.brochure.recommends"), plan: s("share.brochure.plan"),
            reviews: s("share.brochure.reviews"), why: s("share.brochure.why"),
            maps: s("share.brochure.maps"), website: s("share.brochure.website"),
            review: s("share.brochure.review"), order: s("share.brochure.order"),
            booking: s("share.brochure.booking"), ticket: s("share.brochure.ticket"),
            reservation: s("share.brochure.reservation"), more: s("share.brochure.more"),
            footer: s("share.brochure.footer"), people: s("share.brochure.people"),
            budget: s("share.brochure.budget")
        )
    }

    /// Same rule as the web `isSafeHttpsUrl` for what this module needs: https, parseable, has a host.
    static func isSafeHttpsURL(_ raw: String?) -> Bool {
        guard let raw, !raw.isEmpty, let c = URLComponents(string: raw),
              c.scheme?.lowercased() == "https", let host = c.host, !host.isEmpty else { return false }
        return true
    }

    static func pickPlace(_ p: LivePlace) -> SharedPlace {
        var seen = Set<String>()
        let links: [SharedLink] = p.actions.compactMap { a in
            guard sharedActionKinds.contains(a.kind), a.urlKind == "direct",
                  isSafeHttpsURL(a.url), seen.insert(a.url).inserted else { return nil }
            return SharedLink(kind: a.kind, url: a.url, platform: a.platform)
        }
        return SharedPlace(
            name: p.name,
            category: p.categories.first,
            rating: p.rating,
            ratingCount: p.ratingCount,
            address: p.address,
            phone: p.phone,
            openingHours: p.openingHours,
            priceRangeText: p.priceRangeText,
            reasons: p.reasons.filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty },
            links: links,
            image: isSafeHttpsURL(p.image) ? p.image : nil
        )
    }

    private static func linkLabel(_ l: Labels, _ link: SharedLink) -> String {
        let base: String
        switch link.kind {
        case "maps": base = l.maps
        case "website": base = l.website
        case "review": base = l.review
        case "order": base = l.order
        case "booking": base = l.booking
        case "ticket": base = l.ticket
        case "reservation": base = l.reservation
        default: base = link.kind
        }
        if let platform = link.platform,
           !genericPlatforms.contains(platform.trimmingCharacters(in: .whitespaces).lowercased()) {
            return "\(base) (\(platform))"
        }
        return base
    }

    /// Platform names that only restate the link kind — "Website (Official Website)" says nothing twice.
    private static let genericPlatforms: Set<String> = ["website", "official website", "maps", "google maps"]

    private static func fmtCount(_ n: Int, _ lang: String) -> String {
        let f = NumberFormatter()
        f.numberStyle = .decimal
        f.maximumFractionDigits = 0
        f.locale = Locale(identifier: lang.hasPrefix("en") ? "en_US" : "vi_VN")
        return f.string(from: NSNumber(value: n)) ?? String(n)
    }

    /// 4.0 prints as "4", 4.5 as "4.5" — the same as JavaScript's number-to-string.
    private static func fmtRating(_ r: Double) -> String {
        r == r.rounded(.down) ? String(Int(r)) : String(r)
    }

    /// One place as brochure lines — identical layout to the web `placeBlock`.
    static func placeBlock(_ p: SharedPlace, index: Int, lang: String) -> String {
        let l = labels(lang)
        var lines: [String] = ["\(index + 1). \(p.name)"]
        var meta: [String] = []
        if let r = p.rating {
            let count = p.ratingCount.map { " (\(fmtCount($0, lang)) \(l.reviews))" } ?? ""
            meta.append("★ \(fmtRating(r))\(count)")
        }
        if let c = p.category { meta.append(c) }
        if let pr = p.priceRangeText { meta.append(pr) }
        if !meta.isEmpty { lines.append("   " + meta.joined(separator: " · ")) }
        if let a = p.address { lines.append("   📍 \(a)") }
        if let h = p.openingHours { lines.append("   🕐 \(h)") }
        if let ph = p.phone { lines.append("   ☎ \(ph)") }
        if !p.reasons.isEmpty { lines.append("   \(l.why): " + p.reasons.joined(separator: " · ")) }
        for link in p.links { lines.append("   \(linkLabel(l, link)): \(link.url)") }
        return lines.joined(separator: "\n")
    }

    private static func footer(_ l: Labels, _ url: String) -> String {
        var host = url
        if host.hasPrefix("https://") { host.removeFirst("https://".count) }
        else if host.hasPrefix("http://") { host.removeFirst("http://".count) }
        return l.footer.replacingOccurrences(of: "{url}", with: host)
    }

    static func placesBrochure(title: String, places: [SharedPlace], lang: String, url: String) -> String {
        let l = labels(lang)
        var parts: [String] = ["\(l.recommends): \(title)", ""]
        for (i, p) in places.enumerated() { parts.append(placeBlock(p, index: i, lang: lang)); parts.append("") }
        parts.append(footer(l, url))
        return parts.joined(separator: "\n")
    }

    /// The plan brochure, from the plan STRUCTURE.
    ///
    /// The iOS `TappyPlan` model carries days → activities (time, title, description, cost); the
    /// header uses the turn's subject because the model has no `title`. Nothing model-authored
    /// is copied verbatim beyond those structured fields.
    static func planBrochure(_ plan: TappyPlan, title: String, lang: String, url: String) -> String {
        let l = labels(lang)
        var parts: [String] = ["\(l.plan): \(title)", ""]
        for (i, day) in plan.days.enumerated() {
            parts.append(day.title ?? (lang.hasPrefix("en") ? "Day \(day.day ?? i + 1)" : "Ngày \(day.day ?? i + 1)"))
            for it in day.activities ?? [] {
                let head = [it.time, it.title].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: " ")
                if !head.isEmpty { parts.append("  " + head) }
                if let d = it.description, !d.isEmpty { parts.append("     \(d)") }
                if let c = it.cost, !c.isEmpty { parts.append("     \(c)") }
            }
            parts.append("")
        }
        parts.append(footer(l, url))
        return parts.joined(separator: "\n")
    }

    /// Fit within `max` without ever cutting a URL — same strategy as the web `compactBrochure`.
    static func compactBrochure(title: String, places: [SharedPlace], lang: String, url: String, max: Int = inboxMaxBody) -> String {
        let full = placesBrochure(title: title, places: places, lang: lang, url: url)
        if full.count <= max { return full }
        let l = labels(lang)
        let head = "\(l.recommends): \(title)"
        let foot = footer(l, url)
        var kept = places.count
        while kept > 0 {
            let shown = Array(places.prefix(kept))
            let more = places.count - kept
            var body: [String] = [head, ""]
            for (i, p) in shown.enumerated() { body.append(placeBlock(p, index: i, lang: lang) + "\n") }
            if more > 0 { body.append(l.more.replacingOccurrences(of: "{n}", with: String(more)) + "\n") }
            body.append(foot)
            let joined = body.joined(separator: "\n")
            if joined.count <= max { return joined }
            kept -= 1
        }
        guard let p = places.first else { return [head, "", foot].joined(separator: "\n") }
        let slim = SharedPlace(name: p.name, category: p.category, rating: p.rating, ratingCount: p.ratingCount,
                               address: p.address, links: Array(p.links.filter { $0.kind == "maps" }.prefix(1)))
        let more = places.count - 1
        let body = [head, "", placeBlock(slim, index: 0, lang: lang), "",
                    more > 0 ? l.more.replacingOccurrences(of: "{n}", with: String(more)) : "", foot].joined(separator: "\n")
        if body.count <= max { return body }
        return String([head, "", "1. \(p.name)", "", foot].joined(separator: "\n").prefix(max))
    }

    static func buildPlacesArtifact(_ view: PlacesLiveView, title: String, lang: String) -> ShareArtifact {
        let url = TappyShare.canonicalOrigin
        let places = view.items.map(pickPlace)
        let l = labels(lang)
        return ShareArtifact(kind: .places, title: title, subject: "\(l.recommends): \(title)",
                             text: placesBrochure(title: title, places: places, lang: lang, url: url), url: url, places: places)
    }

    static func buildPlanArtifact(_ plan: TappyPlan, title: String, lang: String) -> ShareArtifact {
        let url = TappyShare.canonicalOrigin
        let l = labels(lang)
        return ShareArtifact(kind: .plan, title: title, subject: "\(l.plan): \(title)",
                             text: planBrochure(plan, title: title, lang: lang, url: url), url: url, places: [])
    }

    /// Prose → share text (same rules as the web `proseForShare`): emphasis and headings go, a
    /// markdown link keeps its destination as `label: url` when it passes the guard, images go.
    static func proseForShare(_ text: String) -> String {
        var s = text
        s = s.replacingOccurrences(of: #"!\[[^\]]*\]\([^)]*\)"#, with: "", options: .regularExpression)
        s = s.replacingOccurrences(of: #"\*\*(.*?)\*\*"#, with: "$1", options: .regularExpression)
        s = s.replacingOccurrences(of: #"\*(.*?)\*"#, with: "$1", options: .regularExpression)
        s = s.replacingOccurrences(of: #"#{1,3}\s"#, with: "", options: .regularExpression)
        if let re = try? NSRegularExpression(pattern: #"\[([^\]]+)\]\(([^)\s]+)\)"#) {
            let ns = s as NSString
            var out = ""; var last = 0
            for m in re.matches(in: s, range: NSRange(location: 0, length: ns.length)) {
                out += ns.substring(with: NSRange(location: last, length: m.range.location - last))
                let label = ns.substring(with: m.range(at: 1)), url = ns.substring(with: m.range(at: 2))
                out += isSafeHttpsURL(url) ? "\(label): \(url)" : label
                last = m.range.location + m.range.length
            }
            out += ns.substring(from: last)
            s = out
        }
        if let re = try? NSRegularExpression(pattern: #"(^|\s)(https?://\S+)"#) {
            let ns = s as NSString
            var out = ""; var last = 0
            for m in re.matches(in: s, range: NSRange(location: 0, length: ns.length)) {
                out += ns.substring(with: NSRange(location: last, length: m.range.location - last))
                let pre = ns.substring(with: m.range(at: 1)), url = ns.substring(with: m.range(at: 2))
                out += isSafeHttpsURL(url) ? pre + url : pre
                last = m.range.location + m.range.length
            }
            out += ns.substring(from: last)
            s = out
        }
        s = s.replacingOccurrences(of: #"[ \t]+\n"#, with: "\n", options: .regularExpression)
        s = s.replacingOccurrences(of: #"\n{3,}"#, with: "\n\n", options: .regularExpression)
        return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// A turn with no card and no plan: the prose, under the TappyAI header. Strictly more than before.
    static func buildProseArtifact(subject: String, prose: String) -> ShareArtifact {
        let url = TappyShare.canonicalOrigin
        return ShareArtifact(kind: .places, title: subject, subject: "TappyAI: \(subject)",
                             text: "TappyAI\n\n\(proseForShare(prose))\n\n— TappyAI · tappyai.com", url: url, places: [])
    }

    /// The Inbox/handoff-safe body: same content, bounded, URLs intact.
    static func inboxBody(_ a: ShareArtifact, lang: String) -> String {
        if a.kind == .places, !a.places.isEmpty {
            return compactBrochure(title: a.title, places: a.places, lang: lang, url: a.url)
        }
        if a.text.count <= inboxMaxBody { return a.text }
        let lines = a.text.components(separatedBy: "\n")
        let foot = lines.last ?? ""
        var out: [String] = []
        var len = foot.count + 1
        for line in lines.dropLast() {
            if len + line.count + 1 > inboxMaxBody { break }
            out.append(line); len += line.count + 1
        }
        return (out + [foot]).joined(separator: "\n")
    }
}

File: android/app/src/main/java/com/tappyai/app/share/ShareArtifact.kt
package com.tappyai.app.share

import com.tappyai.app.chat.TappyPlan
import java.net.URI
import java.text.NumberFormat
import java.util.Locale

/**
 * The one canonical TappyAI share artifact — a port of `src/lib/share/shareArtifact.ts`.
 *
 * 🔑 SAME INPUT, SAME TEXT AS THE WEB. The brochure format is deliberately byte-identical to
 * the web builder for the same `PlacesLiveView`, so a recommendation shared from Android and
 * one shared from the web look like they came from the same product — because they did.
 *
 * 🚨 WHITELIST, NEVER BLACKLIST. [SharedPlace] names every field that may leave the app and
 * [pickPlace] copies exactly those. Provenance, evidence types, ranks, entity ids, `_tappy_*`
 * and conversation ids have no field to land in.
 *
 * 🚨 NO `distanceKm`. It is a fact about the SENDER's position (privacy), so it is not in the
 * whitelist and cannot be shared.
 */
data class ShareArtifact(
    val kind: Kind,
    val title: String,
    val subject: String,
    val text: String,
    /**
     * The brand entry point for a recommendation, or — once the share sheet has published a
     * plan — the plan's own canonical page `https://www.tappyai.com/plan/<shareId>`.
     */
    val url: String,
    val places: List<SharedPlace>,
    /**
     * For a plan: the `[TAPPY_PLAN]` block verbatim, the payload `POST /api/plans/share` takes.
     * Null for a recommendation, and null for a plan that arrived without its block (a restored
     * message), which then cannot be published and says so.
     */
    val planJson: String? = null,
    /** True once [url] is the plan's published page rather than the brand entry point. */
    val isPlanLink: Boolean = false,
) {
    enum class Kind { PLACES, PLAN }
}

data class SharedPlace(
    val name: String,
    val category: String? = null,
    val rating: Double? = null,
    val ratingCount: Int? = null,
    val address: String? = null,
    val phone: String? = null,
    val openingHours: String? = null,
    val priceRangeText: String? = null,
    val reasons: List<String> = emptyList(),
    val links: List<SharedLink> = emptyList(),
    val image: String? = null,
)

data class SharedLink(val kind: String, val url: String, val platform: String? = null)

object ShareArtifactBuilder {
    /** chat_messages.body CHECK constraint on the web Inbox; text handoffs use the same bound. */
    const val INBOX_MAX_BODY = 4000

    private val SHARED_ACTION_KINDS = setOf("maps", "website", "review", "order", "booking", "ticket", "reservation")

    private class Labels(
        val recommends: String, val plan: String, val reviews: String, val why: String,
        val maps: String, val website: String, val review: String, val order: String,
        val booking: String, val ticket: String, val reservation: String,
        val more: String, val footer: String, val people: String, val budget: String,
    )

    private val VI = Labels(
        recommends = "TappyAI gợi ý", plan = "Kế hoạch từ TappyAI", reviews = "đánh giá", why = "Vì sao",
        maps = "Bản đồ", website = "Website", review = "Review", order = "Đặt món",
        booking = "Đặt phòng", ticket = "Mua vé", reservation = "Đặt chỗ",
        more = "và {n} địa điểm khác", footer = "Gợi ý bởi TappyAI · {url}", people = "{n} người", budget = "Ngân sách",
    )
    private val EN = Labels(
        recommends = "TappyAI recommends", plan = "A plan from TappyAI", reviews = "reviews", why = "Why",
        maps = "Maps", website = "Website", review = "Review", order = "Order",
        booking = "Book", ticket = "Tickets", reservation = "Reserve",
        more = "and {n} more", footer = "Recommended by TappyAI · {url}", people = "{n} people", budget = "Budget",
    )

    private fun labels(lang: String) = if (lang.startsWith("en")) EN else VI

    /** Same rule as the web `isSafeHttpsUrl` for what this module needs: https, parseable, has a host. */
    fun isSafeHttpsUrl(raw: String?): Boolean {
        if (raw.isNullOrBlank()) return false
        val u = runCatching { URI(raw) }.getOrNull() ?: return false
        return u.scheme.equals("https", ignoreCase = true) && !u.host.isNullOrBlank()
    }

    fun pickPlace(p: LivePlace): SharedPlace {
        val seen = HashSet<String>()
        val links = p.actions.mapNotNull { a ->
            if (a.kind !in SHARED_ACTION_KINDS) return@mapNotNull null
            if (a.urlKind != "direct") return@mapNotNull null
            if (!isSafeHttpsUrl(a.url) || !seen.add(a.url)) return@mapNotNull null
            SharedLink(kind = a.kind, url = a.url, platform = a.platform)
        }
        return SharedPlace(
            name = p.name,
            category = p.categories.firstOrNull(),
            rating = p.rating,
            ratingCount = p.ratingCount,
            address = p.address,
            phone = p.phone,
            openingHours = p.openingHours,
            priceRangeText = p.priceRangeText,
            reasons = p.reasons.filter { it.isNotBlank() },
            links = links,
            image = p.image?.takeIf { isSafeHttpsUrl(it) },
        )
    }

    private fun linkLabel(l: Labels, link: SharedLink): String {
        val base = when (link.kind) {
            "maps" -> l.maps; "website" -> l.website; "review" -> l.review; "order" -> l.order
            "booking" -> l.booking; "ticket" -> l.ticket; "reservation" -> l.reservation
            else -> link.kind
        }
        val platform = link.platform?.takeIf { it.trim().lowercase() !in GENERIC_PLATFORMS }
        return if (platform != null) "$base ($platform)" else base
    }

    /** Platform names that only restate the link kind — "Website (Official Website)" says nothing twice. */
    private val GENERIC_PLATFORMS = setOf("website", "official website", "maps", "google maps")

    private fun fmtCount(n: Int, lang: String): String =
        NumberFormat.getIntegerInstance(if (lang.startsWith("en")) Locale.US else Locale("vi", "VN")).format(n)

    /** One place as brochure lines — identical layout to the web `placeBlock`. */
    fun placeBlock(p: SharedPlace, index: Int, lang: String): String {
        val l = labels(lang)
        val lines = ArrayList<String>()
        lines += "${index + 1}. ${p.name}"
        val meta = ArrayList<String>()
        p.rating?.let { r ->
            val count = p.ratingCount?.let { " (${fmtCount(it, lang)} ${l.reviews})" } ?: ""
            meta += "★ ${fmtRating(r)}$count"
        }
        p.category?.let { meta += it }
        p.priceRangeText?.let { meta += it }
        if (meta.isNotEmpty()) lines += "   ${meta.joinToString(" · ")}"
        p.address?.let { lines += "   📍 $it" }
        p.openingHours?.let { lines += "   🕐 $it" }
        p.phone?.let { lines += "   ☎ $it" }
        if (p.reasons.isNotEmpty()) lines += "   ${l.why}: ${p.reasons.joinToString(" · ")}"
        for (link in p.links) lines += "   ${linkLabel(l, link)}: ${link.url}"
        return lines.joinToString("\n")
    }

    /** 4.0 prints as "4", 4.5 as "4.5" — the same as JavaScript's number-to-string. */
    private fun fmtRating(r: Double): String =
        if (r == Math.floor(r)) r.toInt().toString() else r.toString()

    private fun footer(l: Labels, url: String) = l.footer.replace("{url}", url.removePrefix("https://").removePrefix("http://"))

    fun placesBrochure(title: String, places: List<SharedPlace>, lang: String, url: String): String {
        val l = labels(lang)
        val parts = ArrayList<String>()
        parts += "${l.recommends}: $title"
        parts += ""
        places.forEachIndexed { i, p -> parts += placeBlock(p, i, lang); parts += "" }
        parts += footer(l, url)
        return parts.joinToString("\n")
    }

    /**
     * The plan brochure. Structure from `days[].items[]`; `shareText` may contribute one
     * introductory line only when it is short and carries no URL.
     */
    fun planBrochure(plan: TappyPlan, lang: String, url: String): String {
        val l = labels(lang)
        val parts = ArrayList<String>()
        parts += "${l.plan}: ${plan.title}"
        val intro = plan.shareText?.trim().orEmpty()
        if (intro.isNotEmpty() && intro.length <= 160 && !Regex("https?://", RegexOption.IGNORE_CASE).containsMatchIn(intro)) parts += intro
        val facts = ArrayList<String>()
        plan.people?.let { facts += l.people.replace("{n}", it.toString()) }
        plan.budgetTotal?.let { facts += "${l.budget}: $it" }
        if (facts.isNotEmpty()) parts += facts.joinToString(" · ")
        parts += ""
        for (day in plan.days) {
            parts += day.label
            for (it in day.items) {
                parts += "  " + listOf(it.time, it.emoji, it.name).filter { s -> s.isNotBlank() }.joinToString(" ")
                it.description?.takeIf { d -> d.isNotBlank() }?.let { d -> parts += "     $d" }
                val extra = ArrayList<String>()
                it.price?.takeIf { s -> s.isNotBlank() }?.let { s -> extra += s }
                it.address?.takeIf { s -> s.isNotBlank() }?.let { s -> extra += "📍 $s" }
                if (extra.isNotEmpty()) parts += "     ${extra.joinToString(" · ")}"
                it.mapsLink?.takeIf { s -> isSafeHttpsUrl(s) }?.let { s -> parts += "     ${l.maps}: $s" }
                it.bookingLink?.takeIf { s -> isSafeHttpsUrl(s) }?.let { s -> parts += "     ${l.booking}: $s" }
            }
            parts += ""
        }
        parts += footer(l, url)
        return parts.joinToString("\n")
    }

    /** Fit within [max] without ever cutting a URL — same strategy as the web `compactBrochure`. */
    fun compactBrochure(title: String, places: List<SharedPlace>, lang: String, url: String, max: Int = INBOX_MAX_BODY): String {
        val full = placesBrochure(title, places, lang, url)
        if (full.length <= max) return full
        val l = labels(lang)
        val head = "${l.recommends}: $title"
        val foot = footer(l, url)
        var kept = places.size
        while (kept > 0) {
            val shown = places.take(kept)
            val more = places.size - kept
            val body = buildList {
                add(head); add("")
                shown.forEachIndexed { i, p -> add(placeBlock(p, i, lang) + "\n") }
                if (more > 0) add(l.more.replace("{n}", more.toString()) + "\n")
                add(foot)
            }.joinToString("\n")
            if (body.length <= max) return body
            kept--
        }
        val p = places.first()
        val slim = SharedPlace(
            name = p.name, address = p.address, rating = p.rating, ratingCount = p.ratingCount, category = p.category,
            links = p.links.filter { it.kind == "maps" }.take(1),
        )
        val more = places.size - 1
        val body = listOf(head, "", placeBlock(slim, 0, lang), "", if (more > 0) l.more.replace("{n}", more.toString()) else "", foot).joinToString("\n")
        return if (body.length <= max) body else listOf(head, "", "1. ${p.name}", "", foot).joinToString("\n").take(max)
    }

    fun buildPlacesArtifact(view: PlacesLiveView, title: String, lang: String): ShareArtifact {
        val url = TappyShare.CANONICAL_ORIGIN
        val places = view.items.map(::pickPlace)
        val l = labels(lang)
        return ShareArtifact(
            kind = ShareArtifact.Kind.PLACES, title = title, subject = "${l.recommends}: $title",
            text = placesBrochure(title, places, lang, url), url = url, places = places,
        )
    }

    fun buildPlanArtifact(plan: TappyPlan, lang: String, planJson: String? = null): ShareArtifact {
        val url = TappyShare.CANONICAL_ORIGIN
        val l = labels(lang)
        return ShareArtifact(
            kind = ShareArtifact.Kind.PLAN, title = plan.title, subject = "${l.plan}: ${plan.title}",
            text = planBrochure(plan, lang, url), url = url, places = emptyList(), planJson = planJson,
        )
    }

    /**
     * The plan artifact once the server has published it: the canonical page is the payload.
     *
     * 🔑 THE LINK IS THE BROCHURE. `/plan/<shareId>` renders the real Tappy Plan brochure with
     * the plan's own photos and dynamic OG metadata, so every platform gets the URL — its own
     * preview machinery does the rest — plus one human line naming the plan. Nothing here
     * re-describes the itinerary: that would be a second brochure, and it would drift.
     *
     * Same text on every channel (ACTION_SEND, mailto, the system sheet, the Inbox) and the
     * exact URL on the url handoffs (Facebook sharer, Zalo plugin, Messenger) and on Copy.
     */
    fun planLinkArtifact(base: ShareArtifact, canonicalUrl: String): ShareArtifact {
        require(base.kind == ShareArtifact.Kind.PLAN) { "planLinkArtifact needs a plan artifact" }
        return base.copy(
            url = canonicalUrl,
            text = "${base.subject}\n$canonicalUrl",
            isPlanLink = true,
        )
    }

    /**
     * Prose → share text (same rules as the web `proseForShare`): emphasis and headings go, a
     * markdown link keeps its destination as `label: url` when it passes the guard, images go.
     */
    fun proseForShare(text: String): String = text
        .replace(Regex("!\\[[^\\]]*]\\([^)]*\\)"), "")
        .replace(Regex("\\*\\*(.*?)\\*\\*"), "$1")
        .replace(Regex("\\*(.*?)\\*"), "$1")
        .replace(Regex("#{1,3}\\s"), "")
        .replace(Regex("\\[([^\\]]+)]\\(([^)\\s]+)\\)")) { m -> if (isSafeHttpsUrl(m.groupValues[2])) "${m.groupValues[1]}: ${m.groupValues[2]}" else m.groupValues[1] }
        .replace(Regex("(^|\\s)(https?://\\S+)")) { m -> if (isSafeHttpsUrl(m.groupValues[2])) m.value else m.groupValues[1] }
        .replace(Regex("[ \\t]+\n"), "\n")
        .replace(Regex("\n{3,}"), "\n\n")
        .trim()

    /** A turn with no card and no plan: the prose, under the TappyAI header. Strictly more than before. */
    fun buildProseArtifact(subject: String, prose: String): ShareArtifact {
        val url = TappyShare.CANONICAL_ORIGIN
        return ShareArtifact(
            kind = ShareArtifact.Kind.PLACES, title = subject, subject = "TappyAI: $subject",
            text = "TappyAI\n\n${proseForShare(prose)}\n\n— TappyAI · tappyai.com", url = url, places = emptyList(),
        )
    }

    /** The Inbox/handoff-safe body: same content, bounded, URLs intact. */
    fun inboxBody(a: ShareArtifact, lang: String): String {
        if (a.kind == ShareArtifact.Kind.PLACES && a.places.isNotEmpty()) return compactBrochure(a.title, a.places, lang, a.url)
        if (a.text.length <= INBOX_MAX_BODY) return a.text
        val lines = a.text.split("\n")
        val foot = lines.last()
        val out = ArrayList<String>()
        var len = foot.length + 1
        for (line in lines.dropLast(1)) {
            if (len + line.length + 1 > INBOX_MAX_BODY) break
            out += line; len += line.length + 1
        }
        return (out + foot).joinToString("\n")
    }
}
